ICE when use-ing enum variant after glob-importing same-named enum

[playpen](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=17237a55fd54a2981a6113425362cfce)

Reproduction code:
```rust
mod foo {
    pub enum Foo {
        Foo(i32),
    }
}
use foo::*;
use Foo::Foo;
```

```
thread 'rustc' panicked at 'assertion failed: directive.imported_module.get().is_none()', src/librustc_resolve/resolve_imports.rs:943:21
note: rustc 1.36.0 (a53f9df32 2019-07-03) running on x86_64-unknown-linux-gnu
note: compiler flags: -C codegen-units=1 -C debuginfo=2 --crate-type lib
note: some of the compiler flags provided by cargo are hidden
```

If I re-export `pub use Foo::*` in the module, it works as expected, having enum `Foo` in the type-namespace and variant `Foo` in the value-namespace ([playpen](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=e6a83fcfa7f258d3d190d6bf9cb4144d)).

[The assertion was added](https://github.com/rust-lang/rust/blob/eedf6ce4ef54bb03818ab21d714f1b9f13a6b31c/src/librustc_resolve/resolve_imports.rs#L932) late last year.

@petrochenkov could you take a look at this to see if it is valid for the invariant being checked not to hold in this case? If I comment out the `assert` the output is reasonable:

```
error[E0432]: unresolved import `Foo`
 --> file7.rs:8:5
  |
8 | use Foo::Foo;
  |     ^^^ `Foo` is a variant, not a module

warning: unused import: `foo::*`
 --> file7.rs:7:5
  |
7 | use foo::*;
  |     ^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0432, E0601.
For more information about an error, try `rustc --explain E0432`.
```

--- a/src/librustc_resolve/imports.rs
+++ b/src/librustc_resolve/imports.rs
@@ -874,6 +874,12 @@ impl<'a, 'b> ImportResolver<'a, 'b> {
     /// consolidate multiple unresolved import errors into a single diagnostic.
     fn finalize_import(&mut self, import: &'b Import<'b>) -> Option<UnresolvedImportError> {
         let orig_vis = import.vis.replace(ty::Visibility::Invisible);
+        let orig_blacklisted_binding = match &import.kind {
+            ImportKind::Single { target_bindings, .. } => {
+                Some(mem::replace(&mut self.r.blacklisted_binding, target_bindings[TypeNS].get()))
+            }
+            _ => None,
+        };
         let prev_ambiguity_errors_len = self.r.ambiguity_errors.len();
         let path_res = self.r.resolve_path(
             &import.module_path,
@@ -884,6 +890,9 @@ impl<'a, 'b> ImportResolver<'a, 'b> {
             import.crate_lint(),
         );
         let no_ambiguity = self.r.ambiguity_errors.len() == prev_ambiguity_errors_len;
+        if let Some(orig_blacklisted_binding) = orig_blacklisted_binding {
+            self.r.blacklisted_binding = orig_blacklisted_binding;
+        }
         import.vis.set(orig_vis);
         if let PathResult::Failed { .. } | PathResult::NonModule(..) = path_res {
             // Consider erroneous imports used to avoid duplicate diagnostics.

--- a/src/test/ui/imports/issue-62767.rs
+++ b/src/test/ui/imports/issue-62767.rs
@@ -1,0 +1,15 @@
+// check-pass
+
+mod m {
+    pub enum Same {
+        Same,
+    }
+}
+
+use m::*;
+
+// The variant `Same` introduced by this import is not considered when resolving the prefix
+// `Same::` during import validation (issue #62767).
+use Same::Same;
+
+fn main() {}

